### ISSUE TITLE
Multi-Animal DLC Import Error

### ISSUE THREAD

**Issue by ahabchopra** [sgoldenlab/simba#66]
Hi again! Unfortunately, I seem to have run into a few more errors (that I suspect are closely related). The goal of the SimBA project is to create and track classifiers for multi-animal DLC data which includes 2 pups and 1 adult mouse. Our DLC H5 file contains both labeled multi-animal (for the 2 pups) body parts and unique body parts (for the adult). In order to troubleshoot this error, we have tried to split our H5 file into three separate tests: adult only, infants only, and the original merged dataset. The adult only data works perfectly with SimBA, but I'm finding unique issues with both the infants and the merged data.

Infants Only:
Although I am able to successfully import the video and tracking data for the 2 infants, the load project GUI is completely blank (i.e. only has the tabs on the top, but no buttons within. The error on the terminal is attached below:
![CaptureInfants](https://user-images.githubusercontent.com/43099772/92410759-21aed500-f113-11ea-80f9-cc18c0490dcb.PNG)

Merged Data:
When we include both the multi-animal infants and the adult in one file, SimBA does not allow the import of the H5 file, with an length mismatch error (shown below). On the other hand, when we convert the H5 file to a CSV, it is successfully imported through the initial create project step, but later fails when we try to extract features.
![LengthMismatchError](https://user-images.githubusercontent.com/43099772/92410820-60448f80-f113-11ea-8725-0535bad15ef7.PNG)

Any thoughts would be much appreciated and sorry for the long post!

**Comment by sgoldenlab** [comment]
Hi @ahabchopra!

The short answer: could you try using the development version of SimBA and see if it works?

1. Uninstall the version of SimBA you have - `pip uninstall simba-uw-tf` or `pip uninstall simba-uw-no-tf` 

2. Install the development version `pip install simba-uw-tf-dev` and create or load your project. I inserted a fix for this in this simba version the other day. Let me know how it goes! (this is soon going to become the master version). 


Longer answer: I'm working with a few labs with tracking scenarios just like yours. Originally I did not anticipate that different animals may have different number of tracked body-parts, and therefore `simba-uw-tf` and `simba-uw-no-tf`  versions gives you the error you are seeing. Different number of body-parts and differently named body-parts for different animals require more dynamic menus than earlier versions, and I've tried to fix this in the development version. 

That said, the labs I'm working with that have several pups and a dam have yet to perfect their tracking models so the updates I've made with dynamic menues etc. in SimBA are in anticipation of them getting accurate DLC/SLEAP pose - so I haven't been able to verify it's working well throughout. So, if you report any errors you encounter with the development version here, it would be extremely helpful for me so I can tidy up any functions if needed!

**Comment by ahabchopra** [comment]
Hi @sgoldenlab! Thanks for this thorough explanation! I have just a few follow up questions:
1. When installing the "development version," is there a "without integrated TensorFlow" version, i.e. can I perhaps run `pip install simba-uw-no-tf-dev`?

2. The longer explanation definitely makes sense for the "merged" dataset I described above, but do you know why we would be thrown an error for even the data file with just infants? In that case, the two animals have the same number of labeled body parts (4). 

P.S. Will be sure to update you with any future encountered errors for any small function adjustment 👍

**Comment by sgoldenlab** [comment]
Hi @ahabchopra!

1. I only keep one development version, but(!), even though there is a `tf` in the version name, it does not require tensorflow. If you install version `0.57` with `pip install simba-uw-tf-dev==0.57` is should work. 

2. I tried to recreate the first error you were using using the development version by opening a range of different project with different number of animals and boy-parts but I couldn't. I remember having seen this error before, too, but I seem to have fixed it at some point. Of course let me know if you see it on your end and I will look over it again. 

Thanks

**Comment by ahabchopra** [comment]
Hi @sgoldenlab - successfully installed the dev version 0.57, and happy to report that it seems to be handling the multi-animal 2 infant project well. However, I'm still getting a length mismatch error when I try to introduce the adult to the project (although it seems to be a different # mismatch than the previous error I showed above). Any advice on how to approach this situation - I've attached an image of the terminal below:
![CaptureDev](https://user-images.githubusercontent.com/43099772/92496776-94c55380-f1c6-11ea-9659-0efac5872ee8.PNG)

**Comment by sgoldenlab** [comment]
Hi @ahabchopra - I know this is slightly petty and I am not sure what the documentation says (I may have to update it), but can you stick and underscore between the number and the name of the animal and let me know if it works. I.e., `Single_1` and `Infant_2` and `Infant_3`?

**Comment by ahabchopra** [comment]
Hi @sgoldenlab - definitely not a petty suggestion, but unfortunately it seems to throw the same error about the Length mismatch when I try `single_1 Infant_1 Infant_2.` Not quite sure I understand your edit, but do you mean to say that for the pose_config, I should also put an `_1` after all the single bodyparts, `_2 `and` _3` for the infants?

**Comment by sronilsson** [comment]
Yes, we need a way of tying each bodypart to each individual, I try to be clearer and will get you a screenshot when I get to a computer. Try this:

You should be able to name you animals whatever you wish (single, infant1, infant2 should work, I was wrong above).

Next, for the body parts, add `_1` to the end of each bodypart that belongs to single, add `_2` to the end of each bodypart that belongs to infant1, and `_3` to each bodypart that belongs to infant2. Does that work?

I am aware it's a better way to to this with DeepLabCut data as the individual name is in the multiindex header, but haven't found time to implement it yet.

________________________________

**Comment by ahabchopra** [comment]
@sronilsson - thanks for your suggestion - with this new implementation I was able to successfully extract and label features!

**Comment by sgoldenlab** [comment]
Great @ahabchopra - let me know if anything else comes up

**Comment by ahabchopra** [comment]
Hi @sgoldenlab - sorry to have another issue so soon, but when I move on to the next step (i.e. label behavior), no images appear in the GUI to be labeled, although frames were successfully extracted to the "frames" folder in the project. Moreover, after clicking "Open current video" a seemingly infinite number of mini videos appear on my screen. Not sure if the two errors are related, but just thought I would let you know about the second in case it becomes relevant. Any ideas how to go about this?

**Comment by sgoldenlab** [comment]
Hi @ahabchopra - thanks for reporting this. Let me see if I can replicate it on my end.

**Comment by sgoldenlab** [comment]
@ahabchopra - does it come with an error msg in the terminal?

**Comment by ahabchopra** [comment]
@sgoldenlab - it does! Sorry I didn't include that previously. Here it is below: the first error is shown in the first two lines (both begin with "warning") and actually occurs when I open the video through the "video parameters" tab. No error, however, is shown in the main SimBA GUI. 

The second error (starting with "Traceback") occurs when I try to "show current frame" in the infinite videos that open after I "open current video." However, I don't see an error when I open the label behavior tab with the folder of frames or try to save the Annotations CSV.

![MergeError](https://user-images.githubusercontent.com/43099772/92680178-39937e00-f2f8-11ea-8dbe-ecc12863cdab.PNG)

**Comment by sgoldenlab** [comment]
Thanks @ahabchopra - very helpful. I've tried to recreate the error, I got a similar error to what you where seeing (with frames not displaying in some projects), but unfortunately I couldn't replicate that specific error msg. 

Short explanation, could you update SimBA and try again? `pip install simba-uw-tf==0.58`. SimBA now tries to put a text label on each animal in multi-animal projects (https://github.com/sgoldenlab/simba/blob/master/images/Capturesss.JPG) and these labels could become messed up in the annotator in some tracking scenarios. I've inserted a fix to remove the labels if the error pops up. 

The error you are seeing though, suggest the program can't find the specific frame. In your `project_folder/frames/input`, there should be a directory that has the same name as your video. If you are trying to label a video called `Video1.mp4`, you should have a folder in `project_folder\frames\input\` directory called `Video1` which contains all the frames in your video, starting with `0.png`. 

Also, when you try to open the annotator again in the new SimBA, could you delete the `project_folder\labelling_info.txt` file in your project before giving it a go?

**Comment by ahabchopra** [comment]
Hi @sgoldenlab - will try this fix and let you know how it works! A quick clarification, first, though, do you mean to say I should still install the dev version, i.e., `pip install simba-uw-tf-dev==0.58` or rather just `pip install simba-uw-tf==0.58`. Thanks!

**Comment by sgoldenlab** [comment]
Hi @ahabchopra - sorry it's the dev upgrade, `pip install simba-uw-tf-dev --upgrade`

**Comment by ahabchopra** [comment]
Hey @sgoldenlab - I appreciate you helping me work through these small bugs - I think we're close! Currently, the GUI successfully displays the image (with the 0.58 update). However, it seems to think that both infants are "Infant 1" and the adult is both "single" and "Infant 2" simultaneously. I tried to play with the pose_config, but to no avail. Any way to tell SimBA that the two infants are distinct besides the _2 and _3 labeling in the pose_config? Or will this even be a problem in post processing analysis?

![CaptureLabels](https://user-images.githubusercontent.com/43099772/92820465-fa1c6e80-f397-11ea-8eb9-0494789240b6.PNG)

**Comment by sgoldenlab** [comment]
Almost there @ahabchopra, and thanks for helping me troubleshoot this.

I noticed another potential bug if you use unequal number of body-parts per animal, and I have fixed it. I don't know if that was causing the issue you are seeing, but could you upgrade to 0.59 and see if it works? 

I don't have projects with your tracking data so it's difficult to troubleshoot. If it does not work, could you send me your video + the associated features_extracted file and I will take a stab at it?

And, no, its just a visualization issue, so it does not matter, but it will confuse users so I better handle it. You annotations are still assigned to the specific frames and that's the only real critical part. 

I will try to check in here once daily, and might not be as quick to respond in the upcoming weeks. Thanks gain for your help.

**Comment by ahabchopra** [comment]
Almost there indeed @sgoldenlab! Version 0.59 seems to be working much better. Although I'm getting the same visual error in the GUI I described above, I am annotation as if the labels were correct. While the model seems to run well, I seem to be finding an error during visualization. Specifically, the first generate video "Sklearn visualization" step. I tried to install cmaps and it seems to be from a matlab package, but that fix didn't quite work - any thoughts for this one?

![CaptureCmap](https://user-images.githubusercontent.com/43099772/93145478-665de180-f6ba-11ea-9654-f6a659995a3a.PNG)

**Comment by sgoldenlab** [comment]
Hi @ahabchopra - my bad, I had missed the definitions of the color maps in the plot script. Can you try again, I am up to version 0.61 now, and let me know if it runs?

**Beware**, if you have high resolution videos, visualizations in SimBA can be slow at the moment, but I will look to sort this when time allows.

**Comment by ahabchopra** [comment]
Hi @sgoldenlab! Glad to report that the SimBA program is working well so far! Just having a slight issue with the visualization step, specifically on the "Sklearn visualization." I'm getting an "list index out of range error" for the hullColor array, and my small debugging attempts don't seem to have solved it - any thoughts on this one? Terminal error code shown below:
![sklearn](https://user-images.githubusercontent.com/43099772/94591630-4dc2ff00-0256-11eb-90a6-75d4b9468d87.PNG)

**Comment by sgoldenlab** [comment]
Hi @ahabchopra - first, thanks for reporting this. It looks like there could have been a bug limiting the number of colors generated in situations like yours. 

Could you upgrade SimBA (now version 0.68 - `pip install simba-uw-tf-dev --upgrade`) and report back if it is fixed or not?

**Comment by ahabchopra** [comment]
Fix works great @sgoldenlab! Appreciate all of your help!

**Comment by yana-yuhai** [comment]
Hi @sgoldenlab, I am also having a similar issue where when I try and import the h5 files using H5 (multianimalDLC):

"Importing 0 multi-animal DLC h5 files to the current project
All multi-animal DLC .h5 tracking files ordered and imported into SimBA project
in the chosen workflow file format"

It works when I select CSV (DLC/DeepPoseKit) but using that option leads to issues with Extracting Features.  

I currently have simba-uw-tf-dev and I uninstalled it and tried both versions simba-uw-tf-dev==0.57 and simba-uw-tf-dev==0.58 and did not have any luck. 

Thank you for the help!
